doc/programming_model/attributes_quantization.md

```diff
@@ -73,123 +71,400 @@ with oneDNN static quantization, as long as the quantization
+- **Per-channel zero-points** (`mask=1<<dim`): Different zero-points per
+  channel
+- **Block zero-points** (`mask` with `groups`): Block-wise zero-points
+- **Multi-dimensional zero-points** (`mask=(1<<dim1)+(1<<dim2)`):
+  Independent zero-points across multiple dimensions
+
+~~~cpp
+// Per-tensor zero-point
+attr.set_zero_points(DNNL_ARG_SRC, 0, {}, dnnl::memory::data_type::s32);
+
+// Per-channel zero-points
+attr.set_zero_points(DNNL_ARG_WEIGHTS, 1 << 0, {}, dnnl::memory::data_type::s8);
+
+// Block zero-points
+std::vector<dnnl::memory::dim_t> groups = {64, 1};
+attr.set_zero_points(DNNL_ARG_WEIGHTS, (1 << 0) + (1 << 1), groups,
+                     dnnl::memory::data_type::s32);
+~~~
+
+See examples:
+- [Convolution with Per-output-channel Quantization](#convolution-with-per-output-channel-quantization)
+- [Matmul with Precomputed Reductions and Advanced Quantization](#matmul-with-precomputed-reductions-and-advanced-quantization)
+- @ref inference_int8_matmul_cpp
+- @ref weights_decompression_matmul_cpp
+
+@anchor host-side-scalars-and-zero-points
+### Special Case: Host-side Scalar Scaling Factor and Zero-point
+
+When using the GPU engine and per-tensor quantization,
+host-side scaling factor and zero-point are
+supported to reduce copying of data from host to device.
+A memory object for scaling factor or zero-point value should be created as a
+host-side scalar (see @ref dev_guide_host_side_scalars for details) and passed
+to the primitive execution function.
+
+The host scaling factor or zero-point attributes could also
+be set using the following convenience API:
 
 ~~~cpp
 dnnl::primitive_attr attr;
 attr.set_host_scale(DNNL_ARG_DST,
-           memory::data_type::f32);
+           dnnl::memory::data_type::f32);
 
 attr.set_host_zero_point(DNNL_ARG_DST,
-           memory::data_type::s32);
+           dnnl::memory::data_type::s32);
+~~~
+
+See examples:
+- @ref matmul_with_host_scalar_scale_cpp
+
+### Precomputed Reductions
+
+Precomputed reductions could help optimize performance for Large Language Models (LLM).
+
+When using block-wise zero-points for quantized weights, the library must compute
+reductions over the source tensor during matrix multiplication. This involves
+summing source tensor values across groups along the reduction dimension:
+
+\f[
+\dst_{m,n}=\sum_{g=0}^{G-1}\sum_{k={K\over{G}}g}^{{K\over{G}}(g+1)-1}{\src_{m,k}(\weights_{k,n}-zp_{\weights}(g,n))}=\sum_{k=0}^{K-1}{\src_{m,k}\weights_{k,n}}-\sum_{g=0}^{G-1}zp_{\weights}(g,n)\underbrace{\sum_{k={K\over{G}}g}^{{K\over{G}}(g+1)-1}\src_{m,k}}_{R_{m,g}}
+\f]
+
+where `R` represents the precomputed reductions that can be calculated
+externally when quantizing the source tensor,
+therefore removing the need for the library to compute them at runtime.
+
+The library uses @ref dev_guide_attributes API for setting precomputed reductions.
+The supporting attributes can be found in the documentation for each primitive.
```

Comment:
This sentence would be misleading as there's no such information in matmul - the only one primitive that supports this attribute... If you could take care of that, it would be really cool!

Reply:
ooh, good point, I didn't notice matmul didn't get an update, fixed in https://github.com/uxlfoundation/oneDNN/pull/4068/commits/bd3dd7c86050dab356fa0961b2d848d3d9015fbf :)